ambry-cloud/src/test/java/com/github/ambry/cloud/InMemoryCloudDestinationErrorSimulationTest.java

```diff
@@ -0,0 +1,408 @@
+    responseList.addAll(mockNetworkClient.sendAndPoll(requestList, Collections.emptySet(), 100));
+    requestList.clear();
+    while (responseList.size() < sendCount) {
+      responseList.addAll(mockNetworkClient.sendAndPoll(requestList, Collections.emptySet(), 100));
+    }
+    Assert.assertEquals("Only one request and response", responseList.size(), 1);
+    ResponseInfo responseInfo = responseList.get(0);
+    return responseInfo;
+  }
+
+  /**
+   * Do a put directly to the mock servers. This allows for blobs with malformed properties to be constructed.
+   * @return the blob id
+   * @throws Exception
+   */
+  private BlobId doPut(PartitionId partitionId, ServerErrorCode expectedCode) throws Exception {
+    int blobSize = 4096;
+    // direct put DataBlob
+    BlobType blobType = BlobType.DataBlob;
+    BlobProperties blobProperties = new BlobProperties(blobSize, "serviceId", "memberId", "contentType", false, Utils.Infinite_Time,
+        Utils.getRandomShort(random), Utils.getRandomShort(random), false, null, null, null);
+    byte[] userMetadata = new byte[10];
+    random.nextBytes(userMetadata);
+    byte[] putContent = new byte[blobSize];
+    random.nextBytes(putContent);
+    ByteBuf blobContent = PooledByteBufAllocator.DEFAULT.heapBuffer(blobSize);
+    blobContent.writeBytes(putContent);
+
+    BlobId blobId = new BlobId(routerConfig.routerBlobidCurrentVersion, BlobId.BlobIdType.NATIVE,
+        mockClusterMap.getLocalDatacenterId(), blobProperties.getAccountId(), blobProperties.getContainerId(),
+        partitionId, blobProperties.isEncrypted(),
+        blobType == BlobType.MetadataBlob ? BlobId.BlobDataType.METADATA : BlobId.BlobDataType.DATACHUNK);
+
+    ByteBuffer userMetadataBuf = ByteBuffer.wrap(userMetadata);
+
+    // send to Cloud destinations.
+    PutRequest request =
+        new PutRequest(random.nextInt(), "clientId", blobId, blobProperties, userMetadataBuf.duplicate(),
+            blobContent.retainedDuplicate(), blobContent.readableBytes(), blobType, null);
+
+    RequestInfo requestInfo = new RequestInfo(hostname, port, request, replica, null);
+    ResponseInfo responseInfo = sendAndWaitForResponses(requestInfo);
+    Assert.assertEquals("doPut should succeed.", responseInfo.getError(), null);
+    PutResponse response = PutResponse.readFrom(new NettyByteBufDataInputStream(responseInfo.content()));
+    Assert.assertEquals("The PutResponse is not expected.", expectedCode, response.getError());
+    request.release();
+    blobContent.release();
+
+    return blobId;
+  }
+
+  private BlobId doPut(PartitionId partitionId) throws Exception {
+    return doPut(partitionId, ServerErrorCode.No_Error);
+  }
+
+  @Test
+  public void testPutRequestErrorSimulation() throws Exception {
+    // inject error for cloud colo.
+    cloudDestination.setServerErrorForAllRequests(StoreErrorCodes.IOError);
+
+    BlobId blobId = doPut(partitionId, ServerErrorCode.IO_Error);
```

Comment:
The unit tests look pretty good now, but these methods seem unpolished. Can you add comments to each test function and maybe just remove testUndeleteBlobErrorSimulation?

Reply:
Yes, added the comments. For the testUndeletedBlobErrorSimulation, due to the -1 life version, it threw the exception. So I temporarily disabled it. Right now it works.